src/js/tls.js

```diff
@@ -3,85 +3,110 @@
 require('crypto');
 
 var net = require('net');
+var tcp = require('tcp');
 var util = require('util');
-var EventEmitter = require('events').EventEmitter;
+/** @type tls.TlsWrap */
 var TlsWrap = native.TlsWrap;
 var TLS_CHUNK_MAX_SIZE = require('constants').TLS_CHUNK_MAX_SIZE;
 
-function TLSSocket(socket, opts) {
-  if (!(this instanceof TLSSocket))
-    return new TLSSocket(socket, opts);
-  EventEmitter.call(this);
-
-  var tlsOptions = Object.assign({
-    servername: socket.host || socket.hostname,
-    // rejectUnauthorized: false,
-  }, opts);
+function delegates(proto, target) {
+  if (!(this instanceof delegates)) return new delegates(proto, target);
+  this.proto = proto;
+  this.target = target;
+}
 
-  // TODO: currently certification is disabled
-  tlsOptions.rejectUnauthorized = false;
+delegates.prototype.method = function method(name) {
+  var proto = this.proto;
+  var target = this.target;
 
-  // handle the [ca1,ca2,...]
-  if (Array.isArray(tlsOptions.ca)) {
-    tlsOptions.ca = tlsOptions.ca.join('\n');
-  }
+  proto[name] = function delegation() {
+    return this[target][name].apply(this[target], arguments);
+  };
```

Comment:
Hmm, I don't like this way to delegate the methods from `net` to `tls`, which would name the function to `delegation`, could we just assign it?

Btw, why not `Object.assign` for the delegating functions?

Reply:
I'm afraid there is no way more elegant than this since it's required to access instance `this` in delegated method.
Do you have any suggestion to properly set up a proxy on prototype? 😋